docs/features/component-versioning.md

```diff
@@ -0,0 +1,88 @@
+# Roslyn Component Versioning
+
+This document proposes a scheme that allows the authors of Roslyn Components (Analyzers, Generators etc.) to specify the
+set of features the compiler is required to provide in order to load the specified component.
+
+## Rationale
+
+Today, as an author of a Roslyn component, there is no way to declare a particular version of roslyn that you depend on. For instance
+a Generator that requires the _PostInitialization_ feature will fail to load in compilers that did not support it, but only by
+virtue of the roslyn version it references containing types not available in the older compiler.
+
+With analyzers this was generally less of a problem, as the API shape tended to be more stable, and Syntax is generally additive. Newer
+analyzers could continue to work in older compilers as the code paths referencing new syntax types would not be hit. However, newer APIs, such as those introduced for nullable analysis, still have the load time failure issue.
+
+Although blocked from loading, the current user experience of using a component targeting a newer compiler version is sub-par and opaque as to the action the
+user needs to take to resolve it.
+
+Instead, it would be ideal if the author had a way to specify a particular feature of the compiler that is required; older compilers (at least the
+ones going forward from now on) would be able to detect this feature requirement at load time, and specify a specific warning that this version of the compiler cannot load the component.
+
+## Proposal
+
+We introduce a new Attribute `[RoslynFeature(...)]` that a component author can apply to their component definition. The attribute would take a required string parameter, which is a comma separated list of features the component requires in order to load.
+
+`[RoslynFeature("Generator.PostInitialization, LanguageVersion9")]`
+
+_(__Alternative__: this could directly be an array, rather than separated list?)_
+
+These features would be defined in the compiler, and passed to the component loader. The component loader would check that the requirements are met, and issue a warning when unable to meet them. The requirements would be accessible via API to enable the author to strongly reference the requirement (either an Enum or set of readonly properties).
+
+The exact requirement list is TBD, but would likely include language version and specific component features we introduce going forward. These features could be dynamically calculated by the compiler: for instance, whilst still in preview a feature might only be enabled if the selected language version is also preview.
+
+While we'd allow authors to key on language version, we wouldn't be changing any of the existing functionality in the compiler to use these requirements: it would purely be for making decisions about whether to load the component or not.
```

Comment:
What does it mean for a component to build against the latest Roslyn, but only use features that were present in an older one.  Would you allow that component to load in the older one?  How would you permit the higher assembly ref to be satisfied?  What if the user got it wrong-- `MissingMethodException`?